### ISSUE TITLE
doOnData emmiting value even when error is consumed

### ISSUE THREAD

**Issue by Vanethos** [ReactiveX/rxdart#261]
With the following code:

```test("Observables test", () {
        var asyncMapSubject = PublishSubject<String>();

        asyncMapSubject.stream
          .doOnData((data) {
            print("Do on data triggered: $data");
            throw Error();
        })
        .asyncMap((data) {
          print("asyncmap event: $data");
        })
        .listen((data) {
          print("listen: $data");
        }, onError: (error) {
          print("On error ${error.toString()}");
        });

        asyncMapSubject.add("Bleh");
      });
```

I was expecting that, after the error, both `asyncMap` and `listen` would not be called, since there is no data event (the data event threw an exception)

However, the above code gave the following output:

```
Do on data triggered: Bleh
On error Instance of 'Error'
asyncmap event: Bleh
listen: null
```

I'm using "doOnData" since I'm not trying to map any data, and just checking if the data received is valid.

However, when using "map", I have the correct behaviour, the stream does not call `asynMap` or `listen`...

But the `doOnData` documentation on rxDart states that:

  ```
/// Invokes the given callback function when the stream emits an item. In
  /// other implementations, this is called doOnNext.
```

in the `do.dart` transformer, you have this code:

```
if (onData != null) {
                  try {
                    onData(value);
                  } catch (e, s) {
                    controller.addError(e, s);
                  }
                }
                //..
                controller.add(value);
```

So basically, every error will be handled but then it will be added to the stream again 

Is this intended?

**Comment by brianegan** [comment]
In Dart, error events are not "Terminal" events, meaning you can add multiple errors to a Stream. 

This is not the case with other Rx implementations: An error will terminate the Stream. Therefore, we have to make a judgement call:

  1. `doOnData` errors swallow the event and only the error is emitted
  2. `doOnData` errors are reported, but the event still goes through

IMO, the best we can do is align the behavior to the expectations of our users. It sounds like this behavior was unexpected, so I'd be curious to hear from more folks on what they think!

**Comment by Vanethos** [comment]
I'm biased here, since it's my problem, but isn't Dart's `Stream` behavior closer to what you said in `1.`?

It may also be something that I mixed up, `doOnData` is used more for side effects, so the data in the stream shouldn't be affected by it (in that case we'd use a `map`)

This might be a very specific case. I want to validate, not change, the data that comes into the stream and handle the errors in `onError` if there are any.

**Comment by frankpepermans** [comment]
That should be OK, normally all rx handlers are try catch wrapped, meaning any Error would be caught and added as an ErrorEvent
